From b42dc567a9071719d730393e717ade1db25c1994 Mon Sep 17 00:00:00 2001
Message-Id: <b42dc567a9071719d730393e717ade1db25c1994.1387384437.git.minovotn@redhat.com>
In-Reply-To: <58a5fcf24b76c3cf80d671c215b59a0fef76a5c5.1387384437.git.minovotn@redhat.com>
References: <58a5fcf24b76c3cf80d671c215b59a0fef76a5c5.1387384437.git.minovotn@redhat.com>
From: Fam Zheng <famz@redhat.com>
Date: Tue, 10 Dec 2013 02:06:52 +0100
Subject: [PATCH 2/2] vmdk: Allow read only open of VMDK version 3

RH-Author: Fam Zheng <famz@redhat.com>
Message-id: <1386641212-24945-1-git-send-email-famz@redhat.com>
Patchwork-id: 56099
O-Subject: [RHEL-7 qemu-kvm PATCH] vmdk: Allow read only open of VMDK version 3
Bugzilla: 1007710
RH-Acked-by: Kevin Wolf <kwolf@redhat.com>
RH-Acked-by: Stefan Hajnoczi <stefanha@redhat.com>
RH-Acked-by: Jeffrey Cody <jcody@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1029852
Brew:     https://brewweb.devel.redhat.com/taskinfo?taskID=6698629

Signed-off-by: Fam Zheng <famz@redhat.com>
Signed-off-by: Kevin Wolf <kwolf@redhat.com>
(cherry picked from commit 509d39aa22909c0ed1aabf896865f19c81fb38a1)

Upstream commit doesn't have a verbose commit message, but as the
comment says, VMware KB 2064959 [1] advises that backup software can
read version=3 as version=1. This is important for virt-v2v and helps in
importing VMware VMs from vSphere, since most of the vSphere exported
OVF/OVA's come with version=3.

[1]: http://kb.vmware.com/kb/2064959

Signed-off-by: Fam Zheng <famz@redhat.com>
---
 block/vmdk.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 block/vmdk.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/block/vmdk.c b/block/vmdk.c
index 026b440..7ec6427 100644
--- a/block/vmdk.c
+++ b/block/vmdk.c
@@ -605,13 +605,20 @@ static int vmdk_open_vmdk4(BlockDriverState *bs,
         header = footer.header;
     }
 
-    if (le32_to_cpu(header.version) >= 3) {
+    if (le32_to_cpu(header.version) > 3) {
         char buf[64];
         snprintf(buf, sizeof(buf), "VMDK version %d",
                  le32_to_cpu(header.version));
         qerror_report(QERR_UNKNOWN_BLOCK_FORMAT_FEATURE,
                 bs->device_name, "vmdk", buf);
         return -ENOTSUP;
+    } else if (le32_to_cpu(header.version) == 3 && (flags & BDRV_O_RDWR)) {
+        /* VMware KB 2064959 explains that version 3 added support for
+         * persistent changed block tracking (CBT), and backup software can
+         * read it as version=1 if it doesn't care about the changed area
+         * information. So we are safe to enable read only. */
+        error_setg(errp, "VMDK version 3 must be read only");
+        return -EINVAL;
     }
 
     if (le32_to_cpu(header.num_gtes_per_gt) > 512) {
-- 
1.7.11.7

